operator/builtin/input/aws/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,381 @@
+		tmp := timeLayoutParser(c.logStreamNamePrefix, time.Now())
+		logStreamNamePrefixPtr := aws.String(tmp)
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:               limit,
+			LogGroupName:        logGroupNamePtr,
+			LogStreamNamePrefix: logStreamNamePrefixPtr,
+			StartTime:           startTime,
+			NextToken:           nextTokenPtr,
+		}
+	}
+
+	if c.logStreamNamePrefix != "" {
+		tmp := timeLayoutParser(c.logStreamNamePrefix, time.Now())
+		logStreamNamePrefixPtr := aws.String(tmp)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:               limit,
+			LogGroupName:        logGroupNamePtr,
+			LogStreamNamePrefix: logStreamNamePrefixPtr,
+			StartTime:           startTime,
+		}
+	}
+
+	if len(c.logStreamNames) > 0 && nextToken != "" {
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:          limit,
+			LogGroupName:   logGroupNamePtr,
+			LogStreamNames: c.logStreamNames,
+			StartTime:      startTime,
+			NextToken:      nextTokenPtr,
+		}
+	}
+
+	if len(c.logStreamNames) > 0 {
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:          limit,
+			LogGroupName:   logGroupNamePtr,
+			LogStreamNames: c.logStreamNames,
+			StartTime:      startTime,
+		}
+	}
+
+	if nextToken != "" {
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:        limit,
+			LogGroupName: logGroupNamePtr,
+			StartTime:    startTime,
+			NextToken:    nextTokenPtr,
+		}
+	}
+
+	return cloudwatchlogs.FilterLogEventsInput{
+		Limit:        limit,
+		LogGroupName: logGroupNamePtr,
+		StartTime:    startTime,
+	}
+}
+
+// handleEvent is the handler for a AWS Cloudwatch Logs Filtered Event.
+func (c *CloudwatchInput) handleEvent(ctx context.Context, event *cloudwatchlogs.FilteredLogEvent) error {
+	e := map[string]interface{}{
+		"message":        event.Message,
+		"ingestion_time": event.IngestionTime,
+	}
+	entry, err := c.NewEntry(e)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create new entry from record")
+	}
+
+	entry.AddResourceKey("log_group", c.logGroupName)
+	entry.AddResourceKey("region", c.region)
+	entry.AddResourceKey("log_stream", *event.LogStreamName)
+	entry.AddResourceKey("event_id", *event.EventId)
+	entry.Timestamp = fromUnixMilli(*event.Timestamp)
+
+	// This will check ensures we only persist when needed.
+	// In testing I found that many events would have the same Ingestion Time.
+	// In this case it would write the ingestion time even though it was same.
+	// Added this check to reduce redundant writes.
+	// It also ensures we do not get an older ingestion time then what we already have.
+	if *event.IngestionTime > c.startTime {
+		c.startTime = *event.IngestionTime
+		c.Debugf("Writing start time %d to database", *event.IngestionTime)
+		c.persist.Write(c.logGroupName, *event.IngestionTime)
+	}
+
+	// Write Entry
+	c.Write(ctx, entry)
```

Comment:
This should be called before persisting the checkpoint.

Reply:
Moved as requested